### ISSUE TITLE
Inconsistent InterfaceId for concrete implementation of generic grain interface

### ISSUE THREAD

**Issue by benjaminpetit** [dotnet/orleans#2808]
### Issue

With concrete implementations of generic grain interface, we have a discrepancy between the InterfaceId  value stored in the  `GrainInterfaceMap` and the value used in the generated `GrainReference`

### Example

If we have a grain interface like this:
``` csharp
public interface IGenericGrain<T, U> : IGrainWithIntegerKey { ... }
```

Orleans will generate this `GrainReference`:
```csharp
internal class OrleansCodeGenGenericGrainReference<T, U> : global::Orleans.Runtime.GrainReference, global::UnitTests.GrainInterfaces.IGenericGrain<T, U> {
    ...
    protected override global::System.Int32 InterfaceId
    {
	get { return 1634994201; }
    }
    ...
}
```
Here the value `1634994201` comes from `GrainInterfaceUtils.GetGrainInterfaceId(typeof(IGenericGrain<,>))`

If we have this grain implementation on the silo:
``` csharp
class ConcreteGrainWithGenericInterfaceOfIntFloat : Grain, IGenericGrain<int, float> { ... }
```
at runtime, the silo will construct the `GrainInterfaceMap`, but to get the value of the InterfaceId it will call
```GrainInterfaceUtils.GetGrainInterfaceId(typeof(IGenericGrain<int, float>))``` which will return `1985271655`

So on this silo, in the `GrainInterfaceMap` we will have an entry for `InterfaceId=1985271655`, but the `InvokeMethodRequest` sent by the `GrainReference` will contain `InterfaceId=1634994201` which will NOT be in the `GrainInterfaceMap`

It seems that this is not an issue since we don't use the `InterfaceId` that comes from the `InvokeMethodRequest` (yet!)

~~~I think that we have two ways of fixing this:~~~
- ~~~Always use the "generic" id when adding an entry to the `GrainInterfaceMap`~~~
- ~~~Or compute the real interface id at runtime for `GrainReference` is this is a generic~~~

**Comment by benjaminpetit** [comment]
In summary:

- We use the computed interface id of the **open** generic for the value stored in the generated GrainReference and  the generated MethodInvoker
- We use the computed id of the **closed** generic for the value stored in the GrainInterfaceMap

So far this seems logic:
- All grain implementations (open or close) will use the same GrainReference/MethodInvoker
- We want to use the id of the closed generic in the GrainInterfaceMap since `IMyGrain<string>` isn't the same as `IMyGrain<object>` that is also not the same as `IMyGrain<>`

The issue appears when you declare an open generic grain interface, but you only implement a concrete closed grain interface: if you try to lookup by id (the value from Request.InterfaceId) in the GrainInterfaceMap, it will fail because the map will not contains this Id (it will only contains the Id computed from the closed generic)

**Comment by gabikliot** [comment]
Yes, I think we always knew with @sergeybaykov that some more advanced scenarios of generics don't work correctly. Hope you can nail it now. May the force be with you!

**Comment by sergeybykov** [comment]
Resolved via #2833.
